### ISSUE TITLE
Add InvalidType to analyzer and use it to suppress follow-on errors

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#36697]
Currently, when analyzing erroneous code for which a type can't be determined, the analyzer assumes a type of `dynamic`.  So for example, in the code below, the misspelling of `String` on line (1) causes the type of `s` to be `dynamic`, so there is no follow-on error on lines (2) or (3).

```dart
void f(String _) {}
main() {
  Strqing s = 'foo'; // (1)
  print(s.length); // (2)
  f(s); // (3)
}
```

However, using a type of `dynamic` doesn't suppress all follow-on errors.  For instance, in the example above, if `--no-implicit-casts` is enabled, then `dynamic` is no longer assignable to other types, so a follow-on error appears at (3): `The argument type 'dynamic' can't be assigned to the parameter type 'String'`

Furthermore, follow-on errors that aren't related to assignment are not suppressed; for example, in the following code:

```dart
class Base {
  void f(Strqing x) {} // (1)
}
class Derived extends Base {
  void f(String x) {} // (2)
}
```
the misspelling at (1) causes a follow-on error at (2): `'Derived.f' ('(String) → void') isn't a valid override of 'Base.f' ('(dynamic) → void')`.

We believe we can avoid these problems by introducing a new type `InvalidType` to the analyzer's type system.  It will only be created when analyzing user code for which an error has already been reported; therefore it can be used as a signal that no further related errors need to be reported.  So for example, we can cause `isSubtypeOf` to return `true` when either of its arguments is an `InvalidType`, and that would suppress the follow-on errors described above.

**Comment by srawlins** [comment]
This "P1 high" issue at the build system would be greatly alleviated by an `InvalidType` in the analyzer: https://github.com/dart-lang/build/issues/295

**Comment by srawlins** [comment]
This could also pave a path for de-duplication or improved messaging for unknown default parameter value: https://github.com/dart-lang/sdk/issues/26710

**Comment by srawlins** [comment]
We can also not report `COULD_NOT_INFER` when the types involved include an InvalidType: https://github.com/dart-lang/sdk/issues/35828

**Comment by srawlins** [comment]
Here's another issue which this would unblock: https://github.com/dart-lang/sdk/issues/23037

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/300181

**Comment by scheglov** [comment]
`InvalidType` was added, which fixes the original issue.
Many linked issues were also fixed with additional following CLs.
